dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,277 @@
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+    GDRIVE_USER_CREDENTIALS_DATA = "GDRIVE_USER_CREDENTIALS_DATA"
+    DEFAULT_USER_CREDENTIALS_FILE = ".dvc/tmp/gdrive-user-credentials.json"
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
+        self.gdrive_client_id = self.config.get(
+            Config.SECTION_GDRIVE_CLIENT_ID, None
+        )
+        self.gdrive_client_secret = self.config.get(
+            Config.SECTION_GDRIVE_CLIENT_SECRET, None
+        )
+        if not self.gdrive_client_id or not self.gdrive_client_secret:
+            raise DvcException(
+                "Please specify Google Drive's client id and "
+                "secret in DVC's config. Learn more at "
+                "https://man.dvc.org/remote/add."
+            )
+        self.gdrive_user_credentials_path = self.config.get(
+            Config.SECTION_GDRIVE_USER_CREDENTIALS_FILE,
+            self.DEFAULT_USER_CREDENTIALS_FILE,
+        )
+
+        self.root_id = self.get_path_id(self.path_info, create=True)
+        self.cached_dirs, self.cached_ids = self.cache_root_dirs()
+
+    @on_exception(expo, DvcException, max_tries=8)
+    def execute_request(self, request):
+        try:
+            result = request.execute()
+        except Exception as exception:
+            retry_codes = ["403", "500", "502", "503", "504"]
+            if any(code in str(exception) for code in retry_codes):
+                raise DvcException("Google API request failed")
+            raise
+        return result
+
+    def list_drive_item(self, query):
+        list_request = RequestListFilePaginated(self.drive, query)
+        page_list = self.execute_request(list_request)
+        while page_list:
+            for item in page_list:
+                yield item
+            page_list = self.execute_request(list_request)
+
+    def cache_root_dirs(self):
+        cached_dirs = {}
+        cached_ids = {}
+        for dir1 in self.list_drive_item(
+            "'{}' in parents and trashed=false".format(self.root_id)
+        ):
+            cached_dirs.setdefault(dir1["title"], []).append(dir1["id"])
+            cached_ids[dir1["id"]] = dir1["title"]
+        return cached_dirs, cached_ids
+
+    @cached_property
+    def drive(self):
+        from pydrive.auth import GoogleAuth
+        from pydrive.drive import GoogleDrive
+
+        if os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA):
+            with open(
+                self.gdrive_user_credentials_path, "w"
```

Comment:
we should generate a random tmp file here - to make it atomic (in case someone breaks DVC in the middle, before os remove is called). I think there should not be a default value for the `gdrive_user_credentials_path` at all.

Reply:
> to save the credentials between sessions

Yes, DVC session should know from where to try to load credentials (and to request new one if old were not found)

 > I wonder if we would need an easy mechanism (a flag?) to force the rese

Since now it's possible easy to customize path to credentials file. How about to generate random credentials file name on every run of DVC if `GDRIVE_USER_CREDENTIALS_DATA` is set? Even on failure of credentials file deletion DVC will load expected credentials.

Reply:
@MaxRis there are two different questions (haha, like everything it's not simple with GDrive :) )

- generating `tmp` path to save credentials if `GDRIVE_USER_CREDENTIALS_DATA` is set and I think we are on the same page and it's clear why is it needed, etc.
- (not actually related to the initial comment even) - do we need a mechanism to reset credentials we got from the regular process (when `GDRIVE_USER_CREDENTIALS_DATA` is not set, and we write them to the default location to being able pick them up later)? Let's imagine I see that they are not valid anymore (expired, revoked, whatnot) - how will I know what should I be doing to refresh them?

Reply:
> Let's imagine I see that they are not valid anymore (expired, revoked, whatnot) - how will I know what should I be doing to refresh them?

Good point. Processing of auth exceptions and printing more meaningful error message might will help here. Currently GoogleAuth is configured with `GoogleAuth.DEFAULT_SETTINGS["get_refresh_token"] = True` and expired token will be refreshed automatically, but revoking Google Project access from user's GDrive will cause auth failure for previously used credentials. Haven't tested heavily such scenarios. Will check.

Reply:
Added to the To-Do list as part of issue https://github.com/iterative/dvc/issues/2865